source/gui/guiHelper.py

```diff
@@ -364,18 +364,23 @@ def addDialogDismissButtons(self, buttons, separated=False):
 
 
 def safeAppExit():
-	def ensureMainLoopExited():
-		if wx.GetApp().GetMainLoop():
-			wx.GetApp().ExitMainLoop()
+	"""
+	Ensures the app is exited by the top window being destroyed
+	and that the main loop finally exits as part of that process
+	"""
 
 	if not wx.GetApp().GetMainLoop():
 		log.warning("app trying to exit after already exited")
 		return
 	if not wx.GetApp().GetTopWindow().IsBeingDeleted():
-		# it is important for that to trigger the wx exit process by destroying the top window
+		# it is important for the wxApp exit process to be triggered by destroying the top window
 		# ensure that this is executed on NVDA's main (GUI) thread
 		return wx.CallAfter(wx.GetApp().GetTopWindow().Destroy)
-	# if this is triggered by the top window being destroyed, exit the main loop
+
+	def ensureMainLoopExited():
+		if wx.GetApp().GetMainLoop():
+			wx.GetApp().ExitMainLoop()
+
 	# ensure that this is executed on NVDA's main (GUI) thread
 	wx.CallAfter(ensureMainLoopExited)

```

Comment:
Should both of these (destroy top window, and main loop exit) be queued at once? At the moment it looks like this code relies on safeAppExit being called twice?

Reply:
depending on how `safeAppExit` is initially called, it might be called additionally (and recursively) by the destruction of various dialogs. 

The call for the `safeAppExit` could already be called in the system if the top window has already been destroyed, such as by a close button.

Reply:
If it was only called once, that could result in the top level window being destroyed, but not the main loop being exited. It also seems like this could queue lots of the same actions? Another option could be to set state on the core, which itself handles shutting down the app.

EG:
```python
def safeAppExit():
  from core import shutdownApp
  shutdownApp()
	
...
# core.py
def _ensureMainLoopExited():
	...

def _destroyTopLevelWindow():
	...

doingShutdownLock = threading.Lock()
def shutdownApp(): # I think this probably belongs in core, since that is where the wx App is initialised.
	with doingShutdownLock:
		# Track whether the shutdown process is started, because of wx.CallAfter there is a delay before shutdown state can be queried. We don't want many callers resulting in many queued attempts to shutdown the app. Make starting the process atomic.
		global doingShutdown 
		if not doingShutdown:
			doingShutdown= True
			wx.CallAfter(_ensureMainLoopExited)
			wx.CallAfter(_destroyTopLevelWindow)
		else:
			log.debug("already exited message")
```

Reply:
I've moved this to the MainFrame, as the wx.app mainFrame is initialized in `gui.initialize`. I want to run the new installer system tests on this with better logging to see what behaviour is occurring to hopefully simplify the process.

Reply:
I feel fairly confident this is safe now, given the tests are passing without crashes